agents/vmagent/vmagent.go

```diff
@@ -0,0 +1,169 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package vmagent
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"io/ioutil"
+	"net"
+	"net/http"
+	"net/url"
+	"os"
+	"path"
+	"runtime/pprof"
+	"strconv"
+
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+
+	"github.com/percona/pmm-agent/agents/process"
+	"github.com/percona/pmm-agent/config"
+)
+
+// VMAgent represents VictoriaMetrics agent.
+type VMAgent struct {
+	remoteInsecure   bool
+	remoteUserName   string
+	remotePassword   string
+	binaryPath       string
+	client           *http.Client
+	remoteWriteURL   *url.URL
+	listenURL        *url.URL
+	scrapeConfigPath string
+	lastConfig       []byte
+	tmpDir           string
+	l                *logrus.Entry
+	cancel           context.CancelFunc
+	done             chan struct{}
+}
+
+// NewVMAgent - creates vmagent object and scrape config file.
+func NewVMAgent(cfg *config.Config) (*VMAgent, error) {
+	remoteWriteURL := *cfg.Server.URL()
+	remoteWriteURL.Path = path.Join(remoteWriteURL.Path, "victoriametrics", "api", "v1", "write")
+	remoteWriteURL.User = nil
+	scrapeCfgPath := path.Join(cfg.Paths.TempDir, "vmagent-scrape-config.yaml")
+	scrapeCfg, err := ioutil.ReadFile(scrapeCfgPath) //nolint:gosec
+	if err != nil {
+		if !os.IsNotExist(err) {
+			return nil, errors.Wrapf(err, "failed read vmagent-scrape-config file at %q", scrapeCfgPath)
+		}
+		f, err := os.Create(scrapeCfgPath)
+		if err != nil {
+			return nil, errors.Wrapf(err, "failed to create vmagent-scrape-config file at %q", scrapeCfgPath)
+		}
+		_ = f.Close() //nolint:errcheck
+	}
+
+	return &VMAgent{
+		remoteInsecure: cfg.Server.InsecureTLS,
+		remoteUserName: cfg.Server.Username,
+		remotePassword: cfg.Server.Password,
+		binaryPath:     cfg.Paths.VMAgent,
+		client:         new(http.Client),
+		remoteWriteURL: &remoteWriteURL,
+		listenURL: &url.URL{
+			Host:   net.JoinHostPort("127.0.0.1", strconv.Itoa(int(cfg.Ports.VMAgent))),
+			Scheme: "http",
+		},
+		scrapeConfigPath: scrapeCfgPath,
+		lastConfig:       scrapeCfg,
+		tmpDir:           path.Join(cfg.Paths.TempDir, "vmagent-tmp-dir"),
+		l: logrus.WithFields(logrus.Fields{
+			"component": "agent-process",
+			"agentID":   "vmagent",
+			"type":      "vmagent",
+		}),
+	}, nil
+}
+
+// Start starts vmagent process.
+func (vma *VMAgent) Start(ctx context.Context) {
+	ctx, cancel := context.WithCancel(ctx)
+	pr := &process.Params{Path: vma.binaryPath, Args: vma.buildArgs()}
+	vma.l.Debugf("Starting: %s.", pr)
+	process := process.New(pr, nil, vma.l)
+	go pprof.Do(ctx, pprof.Labels("agentID", "vmagent", "type", "vmagent"), process.Run)
+	done := make(chan struct{})
+	go func() {
+		for status := range process.Changes() {
+			vma.l.Infof("vmanget status changed: %s ", status)
+		}
+		close(done)
+	}()
+	vma.cancel = cancel
+	vma.done = done
+}
+
+// buildArgs returns vmagent process args.
+func (vma *VMAgent) buildArgs() []string {
```

Comment:
maybe it's better to call it just `args`?

Reply:
done